api/src/main/java/io/opencensus/trace/TraceState.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.trace;
+
+import io.opencensus.internal.Utils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Carries tracing-system specific context in a list of key-value pairs. TraceState allows different
+ * vendors propagate additional information and inter-operate with their legacy Id formats.
+ *
+ * <p>Implementation is optimized for a small list of key-value pairs.
+ *
+ * <p>Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+ * and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+ * forward slashes /.
+ *
+ * <p>Value is opaque string up to 256 characters printable ASCII RFC0020 characters (i.e., the
+ * range 0x20 to 0x7E) except comma , and =.
+ *
+ * @since 0.16
+ */
+@Immutable
+public final class TraceState {
+  private static final int KEY_MAX_SIZE = 256;
+  private static final int VALUE_MAX_SIZE = 256;
+  private static final int MAX_KEY_VALUE_PAIRS = 128;
+
+  // Immutable list of key-value pairs.
+  private final List<Entry> entries;
+
+  public static final TraceState EMPTY =
+      new TraceState(Collections.unmodifiableList(new ArrayList<Entry>()));
+
+  private TraceState(List<Entry> entries) {
+    this.entries = entries;
+  }
+
+  /**
+   * Returns the value to which the specified key is mapped, or null if this map contains no mapping
+   * for the key.
+   *
+   * @param key with which the specified value is to be associated
+   * @return the value to which the specified key is mapped, or null if this map contains no mapping
+   *     for the key.
+   */
+  @Nullable
+  public String get(String key) {
+    for (Entry entry : entries) {
+      if (entry.key.equals(key)) {
+        return entry.value;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Returns a {@link List} view of the mappings contained in this {@code TraceState}.
+   *
+   * @return a {@link List} view of the mappings contained in this {@code TraceState}.
+   */
+  public List<Entry> entryList() {
+    return entries;
+  }
+
+  /**
+   * Creates a TraceState by appending the extra entries to the {@code parent} in front of the
+   * key-value pairs list and removing duplicate entries.
+   *
+   * @param parent the parent {@code TraceState}.
+   * @param extraEntries the list of extra entries.
+   * @return a TraceState by appending the extra entries to the {@code parent}.
+   */
+  public static TraceState create(TraceState parent, List<Entry> extraEntries) {
```

Comment:
So if on incoming request I want to read and modify some properties (like read `app-id` of source component and set my own `app-id` for downstream services - will I re-create Tracestate with new entries and set it on `SpanContext`? Also I almost want to make it inside `tags` API as for hierarchical identifiers stored in tracestate you'd need to make an update before making the call to external service somehow.

I guess if you can add comment on the intended use in Javadoc - it will be helpful

Reply:
possible operations on the tracestate seems to be 
- get entry
- add entry
- remove entry
- update entry

And the flow is:
1. copy the parent state 
2. call vendor A, let it check tracestate and add/remove/update entriers
3. call vendor B, etc...

To allow such kind of scenarios (and make them more efficient), I think there should be explicit `copy state`,  `add entry`, `update entry` and `remove entry` methods.

The usage might look like

```
onNewSpan(SpanContext parentContext) {
  childState = parentContext.getTracestate().copy();
  childSpanContext = ...
  childSpanContext.setTraceState(childState);
  for (vendor : vendors) {
    vendor.onNewSpan(childState);
  }
}

...

class MicrosoftAppInsightsExtension : implements Vendor {
  public void onNewSpan(SpanContext spanContext) {
    if (spanContext.getTracestate().get("ms-app-id")) {
      spanContext.getTraceState.update("ms-app-id", "some new value");
    }
  }
}

```

Reply:
So you also suggesting to have a setter for the `tracestate` in `SpanContext`? In this case would it make sense to define `Tracestate` as interface or abstract class? 

(I like it too to make it immutable... so there is less problems with updating it from different child threads)

Reply:
I don't have a strong opinion on how tracestate should be set. SpanContext constructor could create it once and builder could make it immutable. 

My point above was that it's better to have explicit add/update/delete API on a tracestate/tracestate builder than `create(parent, extra items)` with complicated behavior to update/remove entries.

Reply:
I think the intent of API is to keep it immutable. If `add/remove` methods will return the new object - will it be OK. Typically there is only few `tracestate` keys needs to be updated

Reply:
there could be a builder pattern that allows tracestate entries to be changed, and after tracestate is built, it becomes immutable. There is no reason to keep it immutable while invoking vendor callbacks - they can modify it anyway

Reply:
Changed to a builder pattern. This should kind of solve this issue.